File: app/src/models/branch.ts
import { Commit } from './commit'
import { removeRemotePrefix } from '../lib/remove-remote-prefix'
import { ForkedRemotePrefix } from './remote'

// NOTE: The values here matter as they are used to sort
// local and remote branches, Local should come before Remote
export enum BranchType {
  Local = 0,
  Remote = 1,
}

/** The number of commits a revision range is ahead/behind. */
export interface IAheadBehind {
  readonly ahead: number
  readonly behind: number
}

/** The result of comparing two refs in a repository. */
export interface ICompareResult extends IAheadBehind {
  readonly commits: ReadonlyArray<Commit>
}

/** Basic data about a branch, and the branch it's tracking. */
export interface ITrackingBranch {
  readonly ref: string
  readonly sha: string
  readonly upstreamRef: string
  readonly upstreamSha: string
}

/** Basic data about the latest commit on the branch. */
export interface IBranchTip {
  readonly sha: string
}

/** Default rules for where to create a branch from */
export enum StartPoint {
  CurrentBranch = 'CurrentBranch',
  DefaultBranch = 'DefaultBranch',
  Head = 'Head',
  /** Only valid for forks */
  UpstreamDefaultBranch = 'UpstreamDefaultBranch',
}

/** A branch as loaded from Git. */
export class Branch {
  /**
   * A branch as loaded from Git.
   *
   * @param name The short name of the branch. E.g., `main`.
   * @param upstream The remote-prefixed upstream name. E.g., `origin/main`.
   * @param tip Basic information (sha and author) of the latest commit on the branch.
   * @param type The type of branch, e.g., local or remote.
   * @param ref The canonical ref of the branch
   */
  public constructor(
    public readonly name: string,
    public readonly upstream: string | null,
    public readonly tip: IBranchTip,
    public readonly type: BranchType,
    public readonly ref: string
  ) {}

  /** The name of the upstream's remote. */
  public get upstreamRemoteName(): string | null {
    const upstream = this.upstream
    if (!upstream) {
      return null
    }

    const pieces = upstream.match(/(.*?)\/.*/)
    if (!pieces || pieces.length < 2) {
      return null
    }

    return pieces[1]
  }

  /** The name of remote for a remote branch. If local, will return null. */
  public get remoteName(): string | null {
    if (this.type === BranchType.Local) {
      return null
    }

    const pieces = this.ref.match(/^refs\/remotes\/(.*?)\/.*/)
    if (!pieces || pieces.length !== 2) {
      // This shouldn't happen, the remote ref should always be prefixed
      // with refs/remotes
      throw new Error(`Remote branch ref has unexpected format: ${this.ref}`)
    }
    return pieces[1]
  }
  /**
   * The name of the branch's upstream without the remote prefix.
   */
  public get upstreamWithoutRemote(): string | null {
    if (!this.upstream) {
      return null
    }

    return removeRemotePrefix(this.upstream)
  }

  /**
   * The name of the branch without the remote prefix. If the branch is a local
   * branch, this is the same as its `name`.
   */
  public get nameWithoutRemote(): string {
    if (this.type === BranchType.Local) {
      return this.name
    } else {
      const withoutRemote = removeRemotePrefix(this.name)
      return withoutRemote || this.name
    }
  }

  /**
   * Gets a value indicating whether the branch is a remote branch belonging to
   * one of Desktop's automatically created (and pruned) fork remotes. I.e. a
   * remote branch from a branch which starts with `github-desktop-`.
   *
   * We hide branches from our known Desktop for remotes as these are considered
   * plumbing and can add noise to everywhere in the user interface where we
   * display branches as forks will likely contain duplicates of the same ref
   * names
   **/
  public get isDesktopForkRemoteBranch() {
    return (
      this.type === BranchType.Remote &&
      this.name.startsWith(ForkedRemotePrefix)
    )
  }
}

File: GitDesktop/GitDesktop/Models/Branch.swift
import Foundation

/// NOTE: The raw values matter: they are used to sort local before remote.
/// Direct port of `electron/app/src/models/branch.ts`.
public enum BranchType: Int, Codable, Sendable, Comparable {
    case local = 0
    case remote = 1

    public static func < (lhs: BranchType, rhs: BranchType) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

/// The number of commits a revision range is ahead/behind.
/// Port of `IAheadBehind`.
public struct AheadBehind: Codable, Sendable, Equatable {
    public var ahead: Int
    public var behind: Int

    public init(ahead: Int, behind: Int) {
        self.ahead = ahead
        self.behind = behind
    }
}

/// Basic data about the latest commit on the branch. Port of `IBranchTip`.
public struct BranchTip: Codable, Sendable, Equatable {
    public var sha: String

    public init(sha: String) {
        self.sha = sha
    }
}

/// Basic data about a branch and the branch it is tracking. Port of `ITrackingBranch`.
public struct TrackingBranch: Sendable, Equatable {
    public var ref: String
    public var sha: String
    public var upstreamRef: String
    public var upstreamSha: String

    public init(ref: String, sha: String, upstreamRef: String, upstreamSha: String) {
        self.ref = ref
        self.sha = sha
        self.upstreamRef = upstreamRef
        self.upstreamSha = upstreamSha
    }
}

/// Default rules for where to create a branch from. Port of `StartPoint`.
/// `upstreamDefaultBranch` is only meaningful for forks; retained as data.
public enum StartPoint: String, Codable, Sendable {
    case currentBranch = "CurrentBranch"
    case defaultBranch = "DefaultBranch"
    case head = "Head"
    case upstreamDefaultBranch = "UpstreamDefaultBranch"
}

/// A branch as loaded from Git. Port of `Branch` in `branch.ts`.
public struct Branch: Sendable, Equatable, Identifiable {
    /// Short name, e.g. `main` (remote branches keep the `origin/` prefix in `name`).
    public var name: String
    /// Remote-prefixed upstream name, e.g. `origin/main`.
    public var upstream: String?
    public var tip: BranchTip
    public var type: BranchType
    /// Canonical ref, e.g. `refs/heads/main`.
    public var ref: String

    public init(name: String, upstream: String?, tip: BranchTip, type: BranchType, ref: String) {
        self.name = name
        self.upstream = upstream
        self.tip = tip
        self.type = type
        self.ref = ref
    }

    public var id: String { ref }

    /// The name of the upstream's remote.
    public var upstreamRemoteName: String? {
        guard let upstream else { return nil }
        guard let range = upstream.range(of: "^(.*?)/.*", options: .regularExpression) else { return nil }
        _ = range
        // Manual split to mirror `(.*?)\/.*` semantics.
        guard let slash = upstream.firstIndex(of: "/") else { return nil }
        let remote = String(upstream[..<slash])
        return remote.isEmpty ? nil : remote
    }

    /// The name of the remote for a remote branch, nil for local branches.
    public var remoteName: String? {
        guard type == .remote else { return nil }
        let prefix = "refs/remotes/"
        guard ref.hasPrefix(prefix) else { return nil }
        let rest = String(ref.dropFirst(prefix.count))
        guard let slash = rest.firstIndex(of: "/") else { return nil }
        return String(rest[..<slash])
    }

    /// The upstream name without the remote prefix.
    public var upstreamWithoutRemote: String? {
        guard let upstream else { return nil }
        return Branch.removeRemotePrefix(upstream)
    }

    /// The branch name without the remote prefix (local names unchanged).
    public var nameWithoutRemote: String {
        guard type != .local else { return name }
        return Branch.removeRemotePrefix(name) ?? name
    }

    /// Whether this is a remote branch on one of Desktop's auto-created
    /// (`github-desktop-`) fork remotes. Such branches are hidden as plumbing.
    public var isDesktopForkRemoteBranch: Bool {
        type == .remote && name.hasPrefix(forkedRemotePrefix)
    }

    // MARK: - Helpers

    /// Mirrors `lib/remove-remote-prefix.ts`: strips the leading `remote/` segment.
    public static func removeRemotePrefix(_ name: String) -> String? {
        guard let slash = name.firstIndex(of: "/") else { return nil }
        let rest = String(name[name.index(after: slash)...])
        return rest.isEmpty ? nil : rest
    }
}
